### PR TITLE
Log more exception tracebacks, refine some existing error messaging

### PR DESCRIPTION
**Pull request by ShadowJonathan**
Resolves https://github.com/matrix-org/synapse/issues/12662

This PR;
- Adds `exc_info=e`/`exc_info=True` in various areas where an exception being logged would be necessary
  - Optionally while also removing the `%s` string interpolation of the exceptions
- In some places, moves exception objects from being string-interpolated to be explicitly derived for re-raised exceptions (`from e`)
- Refines some miscellaneous error reporting to expose more critical information.

### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#changelog). The entry should:
  - Be a short description of your change which makes sense to users. "Fixed a bug that prevented receiving messages from other servers." instead of "Moved X method from `EventStore` to `EventWorkerStore`.".
  - Use markdown where necessary, mostly for `code blocks`.
  - End with either a period (.) or an exclamation mark (!).
  - Start with a capital letter.
  - Feel free to credit yourself, by adding a sentence "Contributed by @github_username." or "Contributed by [Your Name]." to the end of the entry.
* [x] Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)
* [x] [Code style](https://matrix-org.github.io/synapse/latest/code_style.html) is correct
  (run the [linters](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#run-the-linters))

`Signed-off-by: Jonathan de Jong <jonathan@automatia.nl>`

### CODE REVIEW

File: synapse/config/key.py [hunk 1]
```diff
@@ -447,8 +447,8 @@ def _parse_key_servers(
                 except Exception as e:
                     raise ConfigError(
                         "Unable to parse key %s for server %s in "
-                        "trusted_key_servers: %s" % (key_id, server_name, e)
-                    )
+                        "trusted_key_servers:" % (key_id, server_name)
+                    ) from e
```
Comment (erikjohnston) [inline synapse/config/key.py:451]: I thing `from e` happens automatically in python3?

File: synapse/federation/sender/per_destination_queue.py [hunk 2]
```diff
@@ -352,18 +352,18 @@ async def _transaction_transmission_loop(self) -> None:
...
-            logger.info(e)
+            logger.info(e.msg)
         except HttpResponseException as e:
             logger.warning(
-                "TX [%s] Received %d response to transaction: %s",
+                "TX [%s] Received %d response to transaction:",
                 self._destination,
                 e.code,
-                e,
+                exc_info=e,
```
Comment (erikjohnston) [inline synapse/federation/sender/per_destination_queue.py:361]: So this is the sort of place where it really doesn't make much sense to include a stack trace. as its very common for remote servers to return non-200 responses. 

Plus, the stack trace doesn't tell us anything we don't already know at this stage.

**Review by erikjohnston — COMMENTED** [review]
I think there are some valuable tidy ups in here, but I'm concerned by the sheer number of warnings we're making log a stack trace for. A general rule of thumb: warnings that should have stack traces should be errors. There are times when that is not necessarily the case (like I guess for some of the module API callbacks? But even then).

File: synapse/config/server.py [hunk 3]
```diff
@@ -253,7 +253,7 @@ def read_config(self, config: JsonDict, **kwargs: Any) -> None:
         try:
             parse_and_validate_server_name(self.server_name)
         except ValueError as e:
-            raise ConfigError(str(e))
+            raise ConfigError("Failed to parse/validate server name") from e
```
[the remark this answers is not in the document]
Comment (ShadowJonathan) [reply synapse/config/server.py:256]: Ah, i'll revert it then.

```suggestion
            raise ConfigError(str(e))
```

File: synapse/events/third_party_rules.py [hunk 4]
```diff
@@ -423,7 +426,7 @@ async def check_can_shutdown_room(self, user_id: str, room_id: str) -> bool:
                 raise
             except Exception as e:
                 logger.exception(
-                    "Failed to run module API callback %s: %s", callback, e
+                    "Failed to run module API callback %s:", callback, exc_info=e
```
[the remark this answers is not in the document]
Comment (ShadowJonathan) [reply synapse/events/third_party_rules.py:429]: It doesn't, but same rationale as comment above; to show in the diff where the `e` is going

**Review by richvdh — CHANGES_REQUESTED** [review]
can this be split up into smaller changes please

**Comment by ShadowJonathan** [comment]
Marking as draft, as I need to split it and redo the changes, and that'll probably take a little while.

**Comment by richvdh** [comment]
I'm going to close this PR because it needs substantial changes and it's bitrotting. If you're still interested in doing this work, please open smaller PRs!

**Pull request closed without merging**